internal/controller/application_controller.go

```diff
@@ -163,49 +174,142 @@ func (r *ApplicationPullServiceAccountCreator) getImageRepositoryPullSecretNames
+		},
 	}
 
-	// set serviceAccount ownership to application
-	if err := controllerutil.SetOwnerReference(application, &applicationSA, r.Scheme); err != nil {
-		log.Error(err, "failed to set application as owner of SA", "applicationName", application.Name)
-		return err
+	// Set pull secret ownership to application
+	if err := controllerutil.SetOwnerReference(application, aggregatedSecret, r.Scheme); err != nil {
+		log.Error(err, "failed to set application as owner of aggregated secret", "applicationName", application.Name)
+		return nil, err
 	}
 
-	if err := r.Client.Create(ctx, &applicationSA); err != nil {
-		log.Error(err, "failed to create service account", "serviceAccountName", l.Action, l.ActionAdd)
-		return err
+	foundAggregatedSecret := &corev1.Secret{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: aggregatedSecretName, Namespace: application.Namespace}, foundAggregatedSecret)
+
+	if err != nil && errors.IsNotFound(err) {
+		log.Info("Creating aggregated pull secret.")
+		if err := r.Client.Create(ctx, aggregatedSecret); err != nil {
+			log.Error(err, "failed to create aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionAdd)
+			return nil, err
+		}
+		log.Info("Aggregated pull secret created successfully.")
+		return aggregatedSecret, nil
+	} else if err != nil {
+		log.Error(err, "failed to get aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionView)
+		return nil, err
+	}
+
+	// If secret already exists, check if update is needed
+	if string(foundAggregatedSecret.Data[corev1.DockerConfigJsonKey]) != string(marshaledData) {
+		log.Info("Aggregated pull secret content changed, updating.")
+		foundAggregatedSecret.Data = aggregatedSecret.Data // Update the data
+		// Ensure owner references are maintained if the secret already existed
+		if err := controllerutil.SetOwnerReference(application, foundAggregatedSecret, r.Scheme); err != nil {
+			log.Error(err, "failed to update owner reference for existing aggregated secret", "applicationName", application.Name)
+			return nil, err
+		}
+		if err := r.Client.Update(ctx, foundAggregatedSecret); err != nil {
+			log.Error(err, "failed to update aggregated pull secret", "secretName", aggregatedSecretName, l.Action, l.ActionUpdate)
+			return nil, err
+		}
+		log.Info("Aggregated pull secret updated successfully.")
+	} else {
+		log.Info("Aggregated pull secret content is up-to-date.")
+	}
+
+	return foundAggregatedSecret, nil
+}
+
+// updateServiceAccountWithAggregatedSecret updates the ServiceAccount to include
+// the aggregated pull secret as an imagePullSecret.
+func (r *AggregatedPullSecretCreator) updateServiceAccountWithAggregatedSecret(ctx context.Context, sa *corev1.ServiceAccount, aggregatedSecret *corev1.Secret) error {
+	log := ctrllog.FromContext(ctx)
+
+	// Check if the aggregated pull secret is linked to the SA
+	found := false
+	for _, pullSecret := range sa.ImagePullSecrets {
```

Comment:
you have to add new ApplicationSecret to both sections : Secrets & ImagePullSecrets
of SA

Reply:
Updated